### PR TITLE
Implement the reuse of duplicate constant values

### LINKED ISSUE #42178: RyuJIT: inst_RV_RV_TT doesn't check if constants already exists in the data section
To track the issue noticed in https://github.com/dotnet/runtime/pull/42164/
```csharp
public static float Test(float x)
{
    x = x * 10 * 10 * 10 * 10 * 10;

    x = x / 10 / 10 / 10 / 10 / 10;

    x = x + 10 + 10 + 10 + 10 + 10;

    // other fp operations where op2 is a constant.
    return x;
}
```
### Current codegen:
```asm
; Method Test(float):float
G_M61485_IG01:
       C5F877               vzeroupper 
						;; bbWeight=1    PerfScore 1.00

G_M61485_IG02:
       C5FA590585000000     vmulss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA590581000000     vmulss   xmm0, xmm0, dword ptr [reloc @RWD04]
       C5FA59057D000000     vmulss   xmm0, xmm0, dword ptr [reloc @RWD08]
       C5FA590579000000     vmulss   xmm0, xmm0, dword ptr [reloc @RWD12]
       C5FA590575000000     vmulss   xmm0, xmm0, dword ptr [reloc @RWD16]
       C5FA5E0571000000     vdivss   xmm0, xmm0, dword ptr [reloc @RWD20]
       C5FA5E056D000000     vdivss   xmm0, xmm0, dword ptr [reloc @RWD24]
       C5FA5E0569000000     vdivss   xmm0, xmm0, dword ptr [reloc @RWD28]
       C5FA5E0565000000     vdivss   xmm0, xmm0, dword ptr [reloc @RWD32]
       C5FA5E0561000000     vdivss   xmm0, xmm0, dword ptr [reloc @RWD36]
       C5FA58055D000000     vaddss   xmm0, xmm0, dword ptr [reloc @RWD40]
       C5FA580559000000     vaddss   xmm0, xmm0, dword ptr [reloc @RWD44]
       C5FA580555000000     vaddss   xmm0, xmm0, dword ptr [reloc @RWD48]
       C5FA580551000000     vaddss   xmm0, xmm0, dword ptr [reloc @RWD52]
       C5FA58054D000000     vaddss   xmm0, xmm0, dword ptr [reloc @RWD56]
						;; bbWeight=1    PerfScore 110.00

G_M61485_IG03:
       C3                   ret      
						;; bbWeight=1    PerfScore 1.00
RWD00  dd	41200000h
RWD04  dd	41200000h
RWD08  dd	41200000h
RWD12  dd	41200000h
RWD16  dd	41200000h
RWD20  dd	41200000h
RWD24  dd	41200000h
RWD28  dd	41200000h
RWD32  dd	41200000h
RWD36  dd	41200000h
RWD40  dd	41200000h
RWD44  dd	41200000h
RWD48  dd	41200000h
RWD52  dd	41200000h
RWD56  dd	41200000h
; Total bytes of code: 124
```
^ 10.0 constant is duplicated several times in the data section

### Expected codegen (single constant in the data section):
```asm
; Method Test(float):float
G_M61485_IG01:
       C5F877               vzeroupper 
G_M61485_IG02:
       C5FA590585000000     vmulss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA590581000000     vmulss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA59057D000000     vmulss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA590579000000     vmulss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA590575000000     vmulss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA5E0571000000     vdivss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA5E056D000000     vdivss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA5E0569000000     vdivss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA5E0565000000     vdivss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA5E0561000000     vdivss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA58055D000000     vaddss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA580559000000     vaddss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA580555000000     vaddss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA580551000000     vaddss   xmm0, xmm0, dword ptr [reloc @RWD00]
       C5FA58054D000000     vaddss   xmm0, xmm0, dword ptr [reloc @RWD00]

G_M61485_IG03:
       C3                   ret      
RWD00  dd	41200000h
```

there is a some sort of cache logic for such constants in `genSSE2BitwiseOp` but only for two hard-coded values.

PS: ideally 10.0 constant should be loaded just once to a tmp xmm but it's a different issue I guess.

category:cq
theme:cse
skill-level:intermediate
cost:medium

### PR DESCRIPTION
**Pull request by briansull**
- Duplicate values will now be reused in the constants area
- Added MIN_DATA_ALIGN and MAX_DATA_ALIGN constants
- Added dsDataType to record the type of the constant
- Print the floating point values in the constants area
- Changed 'alignment' and 'cnsSize' parameters to be unsigned instead of UNATIVE_OFFSET
- Added method emitDataGenFind to search and return a duplicate constant value

### CODE REVIEW

**Comment by briansull** [comment]
Fixes #42178

**Comment by briansull** [comment]
Sample diffs:

```
; Assembly listing for method Number:TryParseDouble(ReadOnlySpan`1,int,NumberFormatInfo,byref):bool
; Emitting BLENDED_CODE for X64 CPU with SSE2 - Windows
; ReadyToRun compilation
; optimized code
; rbp based frame
; partially interruptible

G_M10185_IG33:
<       movsd    xmm0, qword ptr [reloc @RWD24]
 >      movsd    xmm0, qword ptr [reloc @RWD00]

G_M10185_IG39:
<       movsd    xmm0, qword ptr [reloc @RWD32]
>       movsd    xmm0, qword ptr [reloc @RWD00]

G_M10185_IG51:
<       movsd    xmm0, qword ptr [reloc @RWD40]
 >      movsd    xmm0, qword ptr [reloc @RWD00]

<RWD00  dq	7FF0000000000000h
<RWD08  dq	FFF0000000000000h
<RWD16  dq	FFF8000000000000h
<RWD24  dq	7FF0000000000000h
<RWD32  dq	FFF8000000000000h
<RWD40  dq	FFF8000000000000h

>RWD00  	dq	7FF0000000000000h		;          inf
>RWD08  	dq	FFF0000000000000h		;         -inf
>RWD16  	dq	FFF8000000000000h		;    -nan(ind)
```

**Comment by briansull** [comment]
It looks like I might have a bug in the pooling that I need to investigate,

**Comment by briansull** [comment]
@BruceForstall @CarolEidt PTAL

File: src/coreclr/src/jit/emit.cpp [hunk 1]
```diff
@@ -5691,6 +5696,42 @@ void emitter::emitDataGenEnd()
...
+/*****************************************************************************
+ * Generates a data section constant
+ *
+ * Parameters:
+ *     cnsAddr  - memory location containing constant value
+ *     cnsSize  - size of constant in bytes
+ *     cnsAlign - alignment of constant in bytes
+ *
+ * Returns constant number as offset into data section.
+ */
```
Comment (BruceForstall) [inline src/coreclr/src/jit/emit.cpp:5708]: This is a copy/paste comment that needs to be updated (to the new comment style, also)

File: src/coreclr/src/jit/emit.cpp [hunk 2]
```diff
@@ -5748,37 +5793,40 @@ CORINFO_FIELD_HANDLE emitter::emitFltOrDblConst(double constValue, emitAttr attr
...
     }
 
     // Access to inline data is 'abstracted' by a special type of static member
     // (produced by eeFindJitDataOffs) which the emitter recognizes as being a reference
     // to constant data, not a real static field.
 
-    UNATIVE_OFFSET cnsSize  = (attr == EA_4BYTE) ? 4 : 8;
-    UNATIVE_OFFSET cnsAlign = cnsSize;
+    unsigned cnsSize  = (attr == EA_4BYTE) ? sizeof(float) : sizeof(double);
+    unsigned cnsAlign = (unsigned)cnsSize;
```
Comment (BruceForstall) [inline src/coreclr/src/jit/emit.cpp:5817]: No cast needed here.

File: src/coreclr/src/jit/emit.cpp [hunk 3]
```diff
@@ -5924,6 +5985,9 @@ void emitter::emitDispDataSec(dataSecDsc* section)
         printf(labelFormat, label);
         offset += data->dsSize;
 
+        // printf(" (DEBUG) data->dsType = %d,dsDataType = %d, data->dsSize = %d\n", data->dsType, data->dsDataType,
```
Comment (BruceForstall) [inline src/coreclr/src/jit/emit.cpp:5988]: Remove this?

File: src/coreclr/src/jit/emit.cpp [hunk 4]
```diff
@@ -5701,17 +5742,21 @@ void emitter::emitDataGenEnd()
  *
  * Returns constant number as offset into data section.
  */
-UNATIVE_OFFSET emitter::emitDataConst(const void* cnsAddr, UNATIVE_OFFSET cnsSize, UNATIVE_OFFSET cnsAlign)
+UNATIVE_OFFSET emitter::emitDataConst(const void* cnsAddr, unsigned cnsSize, unsigned cnsAlign, var_types dataType)
 {
-    UNATIVE_OFFSET cnum = emitDataGenBeg(cnsSize, cnsAlign);
-    emitDataGenData(0, cnsAddr, cnsSize);
-    emitDataGenEnd();
+    UNATIVE_OFFSET cnum = emitDataGenFind(cnsAddr, cnsSize, cnsAlign, dataType);
```
Comment (BruceForstall) [inline src/coreclr/src/jit/emit.cpp:5747]: You've introduced an O(n) search for each constant. Presumably, you could construct a function with a large number of constants where this might show up on a trace. It seems like the usual case would be very few searches, but the usual case is also probably very few matches (duplicates). Would be interesting to see some stats about this across our SPMI collections.

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/src/jit/emit.cpp (hunk 3 above)
Comment (briansull) [reply src/coreclr/src/jit/emit.cpp:5988]: I will remove

File: src/coreclr/src/jit/emit.cpp (hunk 2 above)
Comment (briansull) [reply src/coreclr/src/jit/emit.cpp:5817]: Will fix

File: src/coreclr/src/jit/emit.cpp (hunk 1 above)
Comment (briansull) [reply src/coreclr/src/jit/emit.cpp:5708]: OK

File: src/coreclr/src/jit/emit.cpp (hunk 4 above)
Comment (briansull) [reply src/coreclr/src/jit/emit.cpp:5747]: Yes, I understand that.  There are very few constants used by the JIT, so I didn't think it was worth adding a hash table for this.  I can add a check that counts the total number and asserts (in a checked JIT) when it exceeds a certain limit, (like 10,000 or so)

Comment (BruceForstall) [reply src/coreclr/src/jit/emit.cpp:5747]: I'm not sure it's worth an assert. I suppose the search function could bail out after, say, 100 constants and just always add a new (possibly duplicated) constant after that.

Comment (briansull) [reply src/coreclr/src/jit/emit.cpp:5747]: That is another possibility.  Let me run the tests and SPMI to see if we even have this case today.
Using 10k we would hits this after have 142 unique constants and/or switch tables in a single method.

File: src/coreclr/src/jit/emit.cpp [hunk 5]
```diff
@@ -5515,33 +5515,33 @@ UNATIVE_OFFSET emitter::emitFindOffset(insGroup* ig, unsigned insNum)
...
 
     assert(emitDataSecCur == nullptr);
 
-    // The size must not be zero and must be a multiple of 4 bytes
-    // Additionally, 4 bytes is the minimum alignment that will
+    // The size must not be zero and must be a multiple of MIN_DATA_ALIGN
+    // Additionally, MIN_DATA_ALIGN is the minimum alignment that will
     // actually be used. That is, if the user requests an alignment
     // of 1 or 2, they will get  something that is at least 4-byte
     // aligned. We allow the others since 4 is at least 1/2 and its
```
Comment (CarolEidt) [inline src/coreclr/src/jit/emit.cpp:5529]: You have changed this comment so that it's now using a mix of "4" and MIN_DATA_ALIGN. I'd reword the rest of the comment as well.
```suggestion
    // actually be used. That is, if the user requests an alignment
    // less than MIN_DATA_ALIGN, they will get  something that is at least MIN_DATA_ALIGN
    // aligned. We allow smaller alignment to be specified since it is
```

**Review by CarolEidt — COMMENTED** [review]
Overall LGTM - I'm not sure exactly how the dump looks now, but it seems to be a good improvement - thanks!
I'd agree with Bruce's suggestion to consider a limit on the number of reuse constants.

File: src/coreclr/src/jit/emit.cpp (hunk 4 above)
Comment (briansull) [reply src/coreclr/src/jit/emit.cpp:5760]: I settled on only checking for matches in the first 64 entries and the adding the new value if is doesn't match

```
// If we don't find a match in the first 64, then we just add the new constant
// This prevents an O(n^2) search cost
```

**Review by CarolEidt — APPROVED** [review]
LGTM - thanks!

File: src/coreclr/src/jit/emit.cpp [hunk 6]
```diff
@@ -5515,47 +5515,49 @@ UNATIVE_OFFSET emitter::emitFindOffset(insGroup* ig, unsigned insNum)
...
+    // The size must not be zero and must be a multiple of MIN_DATA_ALIGN
+    // Additionally, MIN_DATA_ALIGN is the minimum alignment that will
     // actually be used. That is, if the user requests an alignment
-    // of 1 or 2, they will get  something that is at least 4-byte
-    // aligned. We allow the others since 4 is at least 1/2 and its
-    // simpler to allow it than to check and block.
-    assert((size != 0) && ((size % 4) == 0));
+    // less than MIN_DATA_ALIGN, they will get  something that is at least
+    // MIN_DATA_ALIGN. We allow smaller alignment to be specified since it is
+    // simpler to allow it than to check and block itr.
```
Comment (tannergooding) [inline src/coreclr/src/jit/emit.cpp:5530]:
```suggestion
    // simpler to allow it than to check and block it.
```

File: src/coreclr/src/jit/emit.cpp [hunk 7]
```diff
@@ -5515,47 +5515,49 @@ UNATIVE_OFFSET emitter::emitFindOffset(insGroup* ig, unsigned insNum)
...
+    //
+    assert((size != 0) && ((size % dataSection::MIN_DATA_ALIGN) == 0));
 
-    // This restricts the alignment to: 1, 2, 4, 8, 16, or 32 bytes
-    // Alignments greater than 32 would require VM support in ICorJitInfo::allocMem
+    // This restricts the alignment to MAX_DATA_ALIGN
+    // Alignments greater than 32 also require VM support in ICorJitInfo::allocMem
```
Comment (tannergooding) [inline src/coreclr/src/jit/emit.cpp:5535]:
```suggestion
    // Alignments greater than MAX_DATA_ALIGN also require VM support in ICorJitInfo::allocMem
```

File: src/coreclr/src/jit/emit.cpp [hunk 8]
```diff
@@ -5691,39 +5697,93 @@ void emitter::emitDataGenEnd()
...
-    emitDataGenEnd();
+    UNATIVE_OFFSET cnum     = INVALID_UNATIVE_OFFSET;
+    unsigned       cmpCount = 0;
+    unsigned       curOffs  = 0;
+    dataSection*   secDesc  = emitConsDsc.dsdList;
+    while (secDesc != nullptr)
+    {
+        // Search the existing secDesc entries
+
+        if ((secDesc->dsSize == cnsSize) && (secDesc->dsDataType == dataType) && ((curOffs % alignment) == 0))
```
Comment (tannergooding) [inline src/coreclr/src/jit/emit.cpp:5721]: Can we not just match the raw integral bits rather than needing to check "dataType"?

For example, if you had a `ulong` constant that was `0x8000_0000_0000_0000` you could also use it for a `double` constant that was `-0.0` as they have the same 64-bit pattern. This would likely be useful for scenarios where you are doing floating-point algorithms that are manipulating the underlying bits.

Likewise, I imagine this might be less efficient for certain SIMD code. If you emit a SIMD constant for `<0.0f, 1.0f, 2.0f, 3.0f>` and also have a floating-point constant for `1.0f` then we will have both emitted separately, rather than just having the latter load from element 1 of the SIMD constant.

These might just be future optimization opportunities, but I wanted to call them out.

File: src/coreclr/src/jit/emit.cpp [hunk 9]
```diff
@@ -5880,19 +5942,33 @@ void emitter::emitOutputDataSec(dataSecDsc* sec, BYTE* dst)
...
                     if ((((i + 1) % 16) == 0) && (i + 1 != dscSize))
                     {
-                        printf("\n  ");
+                        printf("\n\t\t\t\t\t");
                     }
                 }
+                switch (dsc->dsDataType)
+                {
+                    case TYP_FLOAT:
+                        printf(" ; float  %9.6g", (double)*reinterpret_cast<float*>(&dsc->dsCont));
```
Comment (tannergooding) [inline src/coreclr/src/jit/emit.cpp:5958]: Why is this converting to `double` for `printf` rather than just keeping it as `float`?

File: src/coreclr/src/jit/emit.cpp [hunk 10]
```diff
@@ -5991,24 +6067,106 @@ void emitter::emitDispDataSec(dataSecDsc* section)
...
+                        printf("\tdq\t%016llXh", *reinterpret_cast<uint64_t*>(&data->dsCont[i]));
+                        printf("\t; %12.9g", *reinterpret_cast<double*>(&data->dsCont[i]));
+                        i += 8;
+                        break;
+
+                    default:
+                        switch (elemSize)
+                        {
+                            case 1:
+                                printf("\tdb\t%02Xh", *reinterpret_cast<uint16_t*>(&data->dsCont[i]));
```
Comment (tannergooding) [inline src/coreclr/src/jit/emit.cpp:6115]: Shouldn't this be `uint8_t`?

File: src/coreclr/src/jit/emit.cpp (hunk 7 above)
Comment (briansull) [reply src/coreclr/src/jit/emit.cpp:5535]: Actually you have add support for alignments greater than 32 to  ICorJitInfo::allocMem, before you can increase the emitter's MAX_DATA_ALIGN.   If we ever do that we may also move the definitions of MAX/MIN_DATA_ALIGN out of emit.h and into compiler.h

File: src/coreclr/src/jit/emit.cpp (hunk 8 above)
Comment (briansull) [reply src/coreclr/src/jit/emit.cpp:5721]: I could make that change.  It would probably match floating point zero to integer zero frequently.
I also could do `(secDesc->dsSize >= cnsSize) `as well.

File: src/coreclr/src/jit/emit.cpp (hunk 7 above)
Comment (tannergooding) [reply src/coreclr/src/jit/emit.cpp:5535]: Should the comment possibly be moved to be on `MAX_DATA_ALIGN` and indicate something like "this can't be increased without also touching `ICorJitInfo::allocMem`"

File: src/coreclr/src/jit/emit.cpp (hunk 8 above)
Comment (tannergooding) [reply src/coreclr/src/jit/emit.cpp:5721]: > It would probably match floating point zero to integer zero frequently

I think this is probably a rare case. On ARM there is the explicit zero register and on x86 I think we prefer `xor reg, reg` and `xorps reg, reg` as they are both smaller code and elided by the instruction decoder.

File: src/coreclr/src/jit/emit.cpp (hunk 9 above)
Comment (briansull) [reply src/coreclr/src/jit/emit.cpp:5958]: AFAIK printf() only supports the printing of doubles

File: src/coreclr/src/jit/emit.cpp [hunk 11]
```diff
@@ -5880,19 +5942,33 @@ void emitter::emitOutputDataSec(dataSecDsc* sec, BYTE* dst)
...
+                        printf("\n\t\t\t\t\t");
                     }
                 }
+                switch (dsc->dsDataType)
+                {
+                    case TYP_FLOAT:
+                        printf(" ; float  %9.6g", (double)*reinterpret_cast<float*>(&dsc->dsCont));
+                        break;
+                    case TYP_DOUBLE:
+                        printf(" ; double %12.9g", *reinterpret_cast<double*>(&dsc->dsCont));
```
[the remark this answers is not in the document]
Comment (briansull) [reply src/coreclr/src/jit/emit.cpp:5961]: We already have printed the binary value. 
Additionally we print the floating point value as a comment in the Asm output to help the user see what value is being loaded.

File: src/coreclr/src/jit/emit.cpp (hunk 10 above)
Comment (briansull) [reply src/coreclr/src/jit/emit.cpp:6115]: Yes, I will change this

File: src/coreclr/src/jit/emit.cpp (hunk 9 above)
Comment (briansull) [reply src/coreclr/src/jit/emit.cpp:5958]: The C  calling convention only passes doubles, not floats

Comment (tannergooding) [reply src/coreclr/src/jit/emit.cpp:5958]: 👍, I'm definitely not as up to date with what the official standards support.

A simple MSVC console app, however, looks to work and Clang/GCC don't give any warnings on compile.

File: src/coreclr/src/jit/emit.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/coreclr/src/jit/emit.cpp:5961]: 👍, I saw us printing the binary value elsewhere but wasn't sure where in the dump that was relative to this.

It would still be nice if we could ensure that values don't conflict in other places though (IMO).
For example, trimming `double` to 9 digits will result in both of the following looking equal:
* `3.141592653589793 (0x400921FB54442D18)` -- Math.PI
* `3.14159265               (0x400921FB53C8D4F1)`

However, they have 8 million uniquely representable values that exist between them.

**Comment by briansull** [comment]
Update with Tanner's suggestions

**Review by tannergooding — APPROVED** [review]
LGTM!

Thanks for the improvements!

**Comment by briansull** [comment]
@BruceForstall You are still listed as **requested changes** (which I have made)

**Review by BruceForstall — APPROVED** [review]
